fixes and updates to inventory intro (#2416)
ansible/ansible-documentation: docs/docsite/rst/inventory_guide/intro_inventory.rst

How to build your inventory
***************************

> Ansible automates tasks on managed nodes or "hosts" in your infrastructure, using a list or group of lists known as inventory. Ansible composes inventory from one or more 'inventory sources', one of these can be the list of host names you pass at the command line, but most Ansible users create inventory files. Your inventory defines the managed nodes you automate and variables associated to those hosts. You can also define groups, which allow for referencing multiple associated hosts to target for your automation or for bulk defining of variables.

Reviewer: ```suggestion
Ansible automates tasks on managed nodes or "hosts" in your infrastructure, using a list or group of lists known as inventory. Ansible composes inventory from one or more 'inventory sources', one of these can be the list of host names you pass at the command line, but most Ansible users create inventory files. Your inventory defines the managed nodes you automate and the variables associated with those hosts. You can also specify groups, which allow for referencing multiple associated hosts to target for your automation or bulk defining of variables.
```

============================================

You can create your inventory file in one of many formats, depending on the inventory plugins you have.

> The most common formats are INI and YAML (as they are built in), which is what we will mostly discuss in this introduction,

Reviewer: ```suggestion
The most common formats are INI and YAML (as they are built-in), which is what we will mostly discuss in this introduction,
```

> but there many more formats and sources are possible.
> 
> A basic INI ``/etc/ansible/hosts`` might look like this:

Reviewer: ```suggestion
but there are many more formats and sources that are possible.

A basic INI ``/etc/ansible/hosts`` might look like this:
```

Passing multiple inventory sources
==================================

You can target multiple inventory sources (static files, directories, dynamic inventory scripts
or anything supported by inventory plugins) at the same time by specifying multiple inventory sources from the command
line (see below) or by configuration, either setting :envvar:`ANSIBLE_INVENTORY` or in ansible.cfg (:ref:`DEFAULT_HOST_LIST`).
his can be useful when you want to target normally separate environments, like staging and production, at the same time for a specific action.

To target two inventory sources from the command line:

.. code-block:: bash

    ansible-playbook get_logs.yml -i staging -i production

.. _inventory_directory:

Organizing inventory in a directory
===================================

> You can consolidate multiple inventory sources in a single directory. The simplest version of this is a directory with multiple files instead of a single inventory file. A single file gets difficult to maintain when it gets too long. If you have multiple teams and multiple automation projects, having one inventory file per team or project lets everyone easily find the hosts and groups that matter to them. You can also still use the files individually or in subsets, depending on how you configure or call Ansible.
> 
> These can use all formats or plugin configuration (YAML, ini, and so on). In this case your directory is your 'single' inventory source but it agregates the mulitiple sources it finds. There are some direcotories and extensions ignored by default, but you can change this in configuration (:ref:`INVENTORY_IGNORE_PATTERNS` and :ref:`INVENTORY_IGNORE_EXT`).

Reviewer: ```suggestion
You can consolidate multiple inventory sources in a single directory. The simplest version of this is a directory with multiple files instead of a single inventory file. A single file gets difficult to maintain when it gets too long. If you have multiple teams and multiple automation projects, having one inventory file per team or project lets everyone easily find the hosts and groups that matter to them. You can also still use the files individually or in subsets, depending on how you configure or call Ansible.

These can use all formats or plugin configurations (YAML, ini, and so on). In this case, your directory is your 'single' inventory source but it aggregates the multiple sources it finds. There are some directories and extensions ignored by default, but you can change this in configuration (:ref:`INVENTORY_IGNORE_PATTERNS` and :ref:`INVENTORY_IGNORE_EXT`).
```

You can also configure the inventory directory in your ``ansible.cfg`` file. See :ref:`intro_configuration` for more details.

Files will be read and loaded in alphabetically sorted order from the top directory down.

Managing inventory load order
-----------------------------

> Ansible loads inventory sources in the order suplied and will define hosts, groups and variables as it encounters them, adding the ``all`` and ``ungrouped`` at the end, if needed.

Reviewer: ```suggestion
Ansible loads inventory sources in the order supplied and will define hosts, groups and variables as it encounters them, adding the ``all`` and ``ungrouped`` at the end if needed.
```

> Depending on the inventory plugin(s) used you might need to consider rearrainging the order to make sure parent/child defined groups or hosts exist as the plugins involved expect or you might end up with a parsing error.

Reviewer: ```suggestion
Depending on the inventory plugin(s) used you might need to consider rearranging the order to make sure parent/child-defined groups or hosts exist as the plugins involved expect or you might end up with a parsing error.
```

> Variables defined mulitple times are overwritten, last one defined wins.

Reviewer: ```suggestion
Variables defined multiple times are overwritten, last one defined wins.
```

.. _variables_in_inventory:

Adding variables to inventory
=============================

> You can defin variables that relate to a specific host or group in inventory. A simple way to start is adding variables directly to the hosts and groups in a YAML or INI inventory source.

Reviewer: ```suggestion
You can define variables that relate to a specific host or group in inventory. A simple way to start is by adding variables directly to the hosts and groups in a YAML or INI inventory source.
```

Ansible loads inventory sources in the order supplied and will define hosts, groups and variables as it encounters them, adding the ``all`` and ``ungrouped`` at the end if needed.

> Depending on the inventory plugin(s) used you might need to consider rearranging the order to make sure parent/child-defined groups or hosts exist as the plugins involved expect or you might end up with a parsing error. For example, YAML and INI inventory will discard empty groups (no hosts associated) when they end processing each source.

Reviewer: ```suggestion
Depending on the inventory plugin(s) used, you might need to consider rearranging the order to make sure parent/child-defined groups or hosts exist as the plugins involved expect or you might end up with a parsing error. For example, YAML and INI inventory will discard empty groups (no hosts associated) when they end processing each source.
```

    badwolf.example.com:5309

> This is a good way to defind 'Connection variables' (variables used to configure connection, shell, become plugins to enable executing tasks on the host):

Reviewer: ```suggestion
This is a good way to define 'Connection variables' (variables used to configure ``connection``, ``shell``, ``become`` plugins to enable executing tasks on the host):
```

========================

By default, variables are merged/flattened to the specific host before a play is run. This keeps Ansible focused on the Host and Task, so groups do not survive outside of inventory and host matching. By default, Ansible overwrites variables including the ones defined for a group and/or host (see :ref:`DEFAULT_HASH_BEHAVIOUR<DEFAULT_HASH_BEHAVIOUR>`). The order/precedence for inventory entities is (from lowest to highest):

- all group (because it is the 'parent' of all other groups)
- parent group
- child group
- host

By default, Ansible merges groups at the same parent/child level in alphabetical order, and variables from the last group loaded overwrite variables from the previous groups. For example, an ``a_group`` will be merged with ``b_group`` and ``b_group`` vars that match will overwrite the ones in ``a_group``.

.. note:: Ansible merges variables from different sources and applies precedence to some variables over others according to a set of rules. For example, variables that occur higher in an inventory can override variables that occur lower in the inventory. See :ref:`ansible_variable_precedence` for more information.

> You can fine tune this behavior by setting the group variable ``ansible_group_priority``, which overrides the alphabetical sorting for the merge order for groups of the same level (after the parent/child order is resolved). The larger the number, the later it will be merged, giving it higher priority. This variable defaults to ``1`` if not set. For example:

Reviewer: ```suggestion
You can fine-tune this behavior by setting the group variable ``ansible_group_priority``, which overrides the alphabetical sorting for the merge order for groups of the same level (after the parent/child order is resolved). The larger the number, the later it will be merged, giving it higher priority. This variable defaults to ``1`` if not set. For example:
```
